Add Flight mode column to the CSV file

danarrib opened the pull request:
Add a text column "simplifiedMode" to the CSV file.

It's a 4-character active flight mode indicator, almost identical to the INAV Flight Mode OSD element. 

Possible values are:
* `!FS!`
* `MANU`
* `RTH`
* `A+PH`
* `3CRS`
* `CRS`
* `PH`
* `AH`
* `WP`
* `ANGL`
* `HOR`

It'll be useful to display the active flight mode on a overlay tool that I'm planning to publish soon (a command-line dashware-like tool)

![image](https://user-images.githubusercontent.com/17026744/101648535-ad891380-3a18-11eb-9c6d-d695f8da3b1c.png)

![image](https://user-images.githubusercontent.com/17026744/101649084-415adf80-3a19-11eb-96c4-e61c83ca252b.png)

Also, this change:
* Removes the `--dashware` command line option.
* Makes the fields `Distance (m)`, `homeDirection`, `mAhPerKm`, `cumulativeTripDistance` and `azimuth` available on the CSV file when the `--merge-gps` command line option is used.
* Makes the fields `rssi (%)` and `Throttle (%)` always available, since this fields are not related to GPS data.

stronnag commented:
Please fix the following compilation warning:
```
%% blackbox_decode.c
.//src/blackbox_decode.c:104:23: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |  NAV_CRUISE_MODE = (1 << 36),
      |                       ^~
.//src/blackbox_decode.c:104:20: warning: enumerator value for ‘NAV_CRUISE_MODE’ is not an integer constant expression [-Wpedantic]
  104 |  NAV_CRUISE_MODE = (1 << 36),
      |                    ^
```
The ISO C standard defines that an enum is of width int. We prefer to be both warning free and standard compliant.

Where do the values after  MANUAL_MODE come from (they are off by one compared to firmware `src/main/fc/rc_modes.h`

So we have the situation with a log including a waypoint flight, we have:
*  flightModeFlags (flags) : `ARM|NAVWP` (correct)
* navState 17 == `nav_state_waypoint_in_progress` (correct)
* the new `flightMode` column states `ACRO` (which is incorrect)

It would also be preferable if the new column as not called `flightMode` as it's a gross simplification of what the firmware thinks is the flightMode, perhaps `simplifiedMode` or `displayedMode` or `OsdMode` would be better and less confusing.

stronnag commented:
As discussed on TG previously, the actual blackbox variable that is stored in the BBL is only 32 bits wide, so defining non-standard enums beyond (1<<31) is pointless (as well as non-standard).

danarrib commented:
Thanks the feedback Jonathan. I just fixed the problems.
Now the flight mode flags are correctly defined. I removed the unnecessary ones (the ones that doesn't affect the displayed text).
I also did some minor fixes... The columns "rssi (%)" and "Throttle (%)" now are added to resulting file regardless of "--merge-gps" option, since this values are not GPS dependent.

stronnag commented:
Daniel, much better. May I suggest a couple more small improvements:
* Use the BBL `failsafePhase` field to detect failsafe (via defnition in `inav/src/main/flight/failsafe.h` and how it's handled here in `src/parser.c`. This will then work for F/S cases other than RTH
* Use #define or enum for `navStates`. This makes it easier to follow /maintain the code vice magic numbers. 
* Consider then if you want to use the CRS_2D values from `inav/src/main/navigation/navigation_private.h` /  `navigationPersistentId_e`, I don't think your current test for 2D is valid.

stronnag commented:
We also still have a problem with home range and bearing (well, bearing at least). 
Here's what should happen (home is also c. W (278°) of the launch point).
![Screenshot-20201210153400-625x192](https://user-images.githubusercontent.com/158229/101793279-86584200-3afd-11eb-849e-6a17e2da5bdb.png)

The log says it's NE (52°) --- not even 180°off.
![Screenshot-20201210154103-321x163](https://user-images.githubusercontent.com/158229/101794086-683f1180-3afe-11eb-9681-082f16117034.png)

Later on in the log, the bearing is negative. This should not happen. 
![Screenshot-20201210154202-383x446](https://user-images.githubusercontent.com/158229/101794205-89076700-3afe-11eb-8be6-ccf58568e4a3.png)

stronnag commented:
So part of the problem is nomenclature. What's called homeDirection is relative bearing and azimuth is the reciprocal bearing (and incorrect).

stronnag commented:
Daniel, I've been playing around with another approach to the geodetics that for my examples gives somewhat (IMO) resutls, such that especially for flights involving lots of turns, it gets more accurate total distances and energy usage.

Let me know when you've got as far as you're content with your current PR, and I'll rebase an experimental branch for you to try on what you've done so far plus my experiment.

danarrib commented:
> Daniel, much better. May I suggest a couple more small improvements:
> 
> * Use the BBL `failsafePhase` field to detect failsafe (via defnition in `inav/src/main/flight/failsafe.h` and how it's handled here in `src/parser.c`. This will then work for F/S cases other than RTH
> * Use #define or enum for `navStates`. This makes it easier to follow /maintain the code vice magic numbers.
> * Consider then if you want to use the CRS_2D values from `inav/src/main/navigation/navigation_private.h` /  `navigationPersistentId_e`, I don't think your current test for 2D is valid.

Hello Jonathan, thanks again for your valuable feedback. I fixed this points and I'll add it too, it's already on the branch.

Regarding the problems with the homeDirection and azimuth, I can reproduce using the two log files you provided, but with my log file everything works as expected... Today was a busy day and I could not give this problem the needed attention.

Let me know if you'll take a look on this or if you like I try to fix.

stronnag commented:
Daniel, tomorrow I'll merge your PR "as is" and rebase my experiential geodetics on top of it. Then you can test that branch on your files and maybe we reach a consensus.

stronnag commented:
Daniel, please try #24, in particular whether its idea of relative home direction (clockwise from the vehicle) is what you expect.
For some of my aerobatic, range < 300m flights, #24 provides an accurate measure of distance, where #23 could be 50% out (i.e. cumulative distance ~ 50% of the actual distance).
